### DISCUSSION TITLE
Example of per request i18n with express/apollo?

### DISCUSSION THREAD

**Discussion by sleepycat** [lingui/js-lingui#954] [Q&A]
I'm trying to get per request i18n for a GraphQL api, and I'm hoping for a sanity check.

Apollo Server let's you make a custom context object per request. I'm wanting to create an i18n object per request and pass it in like this (assuming it has some locale attached like what [express-request-language](https://www.npmjs.com/package/express-request-language#usage) provides).

```js
import { setupI18n } from "@lingui/core"

const server = new ApolloServer({
  typeDefs,
  resolvers,
  context: (req) => ({
    i18n: setupI18n({locale: req.language})
  })
}));
```
That context object (and it's i18n property) is then available to all the resolvers involved in that request, dependency injection style.

I want lingui to be able to extract the strings I'm using in the resolvers but it seems that the `t` macro refers to a global object, so I think `defineMessage` gives me what I'm looking for.

```js
// Example resolver
import { defineMessage } from "@lingui/macro"

defineMessage({ message: 'round' })
defineMessage({ message: 'square' })

(parent, args, context, info) => {
 const thing = getSomethingFromTheDB()
 return {
    sku: thing._id,
    shape: context.i18n._(thing.shape),
  }
}
```
Is this the best way to achieve what I'm after?

**Comment by semoal** [comment]
Sorry for the late response mate, been busy with Rematch, my actual job etc.
I'm sure that you found which method archive your needs, in my recommendation it's to follow the second form, use an entire file of translation (if you want) with defineMessage and use it like that. That's the best way.
